casper/src/main/resources/RevVault.rho

```diff
@@ -0,0 +1,154 @@
+            for (retCh <- createGenesisVault) {
+
+              _makeVault!(ownerPubKeyBytes, initialAmount, *retCh)
+            }
+          }
+        } |
+
+        contract RevVault(@"findOrCreate", @ownerPubKeyBytes, ret) = {
+
+          new createLockedVault in {
+            _getOrCreate!(ownerPubKeyBytes, *createLockedVault, *ret) |
+            for (retCh <- createLockedVault) {
+
+              new vaultCh, lock in {
+                _makeVault!(ownerPubKeyBytes, 0, *vaultCh) |
+                @Either!("map <-", *vaultCh, *lock, *retCh) |
+                for (@vault, retCh <- lock) {
+                  @Lockbox!("lock", ownerPubKeyBytes, vault, *retCh)
+                }
+              }
+            }
+          }
+        } |
+
+        contract _getOrCreate(@key, constructor, ret) = {
+          for (@vaultMap <- vaultMapStore) {
+            match vaultMap.get(key) {
+              Nil => {
+                new resCh in {
+                  constructor!(*resCh) |
+                  for (@eitherVault <- resCh) {
+                    match eitherVault {
+                      ("Right", vault) => vaultMapStore!(vaultMap.set(key, vault))
+                      ("Left",  _    ) => vaultMapStore!(vaultMap)
+                    } |
+                    ret!(eitherVault)
+                  }
+                }
+              }
+              value => {
+                vaultMapStore!(vaultMap) | ret!(("Right", value))
+              }
+            }
+          }
+        } |
+
+        contract _makeVault(@ownerPubKeyBytes, @initialAmount, ret) = {
+
+          new revAddrCh, eitherRevAddrCh, purseCh, eitherPurseCh in {
+            RevAddress!("fromPublicKey", ownerPubKeyBytes, *revAddrCh) |
+            @Either!("fromNillable <-", *revAddrCh, "Invalid public key", *eitherRevAddrCh) |
+            mint!("makePurse", initialAmount, *purseCh) |
+            @Either!("fromNillable <-", *purseCh, "Couldn't create purse", *eitherPurseCh) |
+            @Either!("map2 <-", *eitherRevAddrCh, *eitherPurseCh, *_newVault, *ret)
+          }
+
+        } |
+
+        contract _newVault(@ownRevAddress, purse, ret) = {
+          new revVault in {
+
+            ret!(bundle+{*revVault}) |
+
+            contract revVault(@"balance", ret) = {
+              purse!("getBalance", *ret)
+            } |
+
+            contract revVault(@"transfer", @revAddress, @amount, ret) = {
+
+              new
+                revAddressValid, revAddressValidEither, amountNonNegative, split,
+                eitherPurseCh, purseReadyCh, doDeposit
+              in {
+                RevAddress!("validate", revAddress, *revAddressValid) |
+                @Either!("fromNillableError <-", *revAddressValid, *revAddressValidEither) |
+                @Either!("fromBoolean", amount >= 0, "Amount must be non-negative", *amountNonNegative) |
+                @Either!("flatMap <-", *amountNonNegative, *split, *eitherPurseCh) |
+                for (_, ret <- split) {
```

Comment:
This is consuming two consecutive values from the split channel (not a Tuple) while `Either.flatMap` sends a tuple to the `f` parameter.

Why is this working? I'm expecting it to block while waiting for `ret` parameter. 

Am I missing something?

Reply:
afaiu it does an arity-2 receive, complementary to an arity-2 send done in https://github.com/rchain/rchain/pull/2268/files#diff-ca522a4895f4cbde9cde7606fe2bc704R131. Notice none of them are sending/receiving a tuple directly (which would be arity-1). Rather, both of them send/receive two values, the first being a 2-tuple (the either) and the second being the return channel

Reply:
Also, AFAIK there's no such ting as 'consecutive values in a channel', as channel receive/send order is non-deterministic